File: applets/asc/counter/cross_counter/assembly/counter.ts
import { Runtime } from '@weilliptic/weil-sdk-asc/assembly/runtime'
import { Result } from '@weilliptic/weil-sdk-asc/assembly/result'
import { Box } from '@weilliptic/weil-sdk-asc/assembly/primitives'
import { WeilError } from '@weilliptic/weil-sdk-asc/assembly/error'

@json
export class CrossCounterContractState {
  constructor() {}

  fetch_counter_from(contractId: string): Result<Box<u64>, WeilError> {
    return Runtime.callContract<u64>(contractId, 'get_count')
  }

  increment_counter_of(contractId: string): Result<Box<bool>, WeilError> {
    return Runtime.callContract<bool>(contractId, 'increment')
  }
}

@json
export class CrossContractArgs {
  contract_id: string

  constructor() {
    this.contract_id = ''
  }
}

File: applets/go/counter/cross_counter/contract/counter.go
package contract

import (
	"github.com/weilliptic-public/wadk/adk/go/weil_go/runtime"
)

type Counter struct {
	ContractId string `json:"contract_id"`
}

func newCounter(contractId string) *Counter {
	return &Counter{
		contractId,
	}
}

func (c *Counter) getCount() (*uint32, error) {
	resp, err := runtime.CallContract[uint32](c.ContractId, "get_count", "")

	if err != nil {
		return nil, err
	} else {
		return resp, nil
	}
}

func (c *Counter) increment() error {
	_, err := runtime.CallContract[interface{}](c.ContractId, "increment", "")

	if err != nil {
		return err
	} else {
		return nil
	}
}
